File: problems/0001-two-sum/python/solution.py
"""LeetCode Problem 1: Two Sum solution."""

from typing import List


class Solution:
    """Solution class for Two Sum problem."""

    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        Find two numbers that add up to target.

        Args:
            nums: List of integers.
            target: Target sum.

        Returns:
            List of two indices whose values sum to target.
        """
        seen = {}
        for i, num in enumerate(nums):
            complement = target - num
            if complement in seen:
                return [seen[complement], i]
            seen[num] = i
        return []

File: problems/0001-two-sum/cpp/solution.cpp
#include <unordered_map>
#include <vector>

using namespace std;

/**
 * @brief 1. Two Sum
 * Given an array of integers nums and an integer target,
 * return indices of the two numbers such that they add up to target.
 * You may assume that each input would have exactly one solution.
 *
 * Time Complexity:  O(n)
 * Space Complexity: O(n)
 */
class Solution {
 public:
  vector<int> twoSum(vector<int>& nums, int target) {
    unordered_map<int, int> seen;  // value -> index

    for (size_t i = 0; i < nums.size(); ++i) {
      int complement = target - nums[i];
      if (seen.find(complement) != seen.end()) {
        return {seen[complement], static_cast<int>(i)};
      }
      seen[nums[i]] = static_cast<int>(i);
    }
    return {};
  }
};
